### ISSUE TITLE
StackOverflow exception when resolving TypeReference SqlDbType

### ISSUE THREAD

**Issue by Cjewett** [coverlet-coverage/coverlet#880]
Hello,

I'm attempting to use the global tool to add code coverage to our code base. It's a mix of .NET Framework 4.6.1 and .NET Standard 2.0.0 as far as unit tests are concerned, and the .NET Framework projects are not SDK-style. For the purposes of this test I did move them to SDK-style projects for testing but I saw this issue both with and without SDK-style .NET Framework projects. Unfortunately I can't share sample code because the libraries being used are proprietary and I'm not sure which part is failing to replicate it in a smaller solution. 

We have a list of projects:
.NET Framework Unit Test 1 
.NET Framework Library 20
.NET Framework Library 21 
.NET Standard Library 31 
.NET Standard Library 32 
.NET Standard Library 33

Reference tree that works:
- NET Framework Unit Test 1 
   - NET Framework Library 21

Reference tree that breaks and I'm trying to solve:
- NET Framework Unit Test 1 
   - NET Framework Library 20
      - NET Standard Library 31
         - NET Standard Library 33
      - NET Standard Library 32
      - NET Standard Library 33
   - NET Framework Library 21

When I build the solution I go to .NET Framework Unit Test 1's output folder and I see all of the libraries DLLs and their PDBs. I check other output folders and the size of the PDBs match. When I run code coverage it seems to run into warnings and hangs.

`coverlet C:\Software\Tests\UnitTests\CoverletTestingLibraryTests\bin\Debug\net461\CoverletTestingLibraryTests.dll --target "dotnet" --targetargs "vstest C:\Software\Tests\UnitTests\CoverletTestingLibraryTests\bin\Debug\net461\CoverletTestingLibraryTests.dll" --format cobertura --output "C:/Software"`

The warning message I get is: "BadImageFormatException during MetadataReaderProvider.FromPortablePdbStream in InstrumentationHelper.PortablePdbHasLocalSource, unable to check if module has got local source." but it is not clear to me which PDB it is complaining about. These PDBs also aren't portable ones, they are specified as full PDBs in the csproj.

Here is the full log up to the hang: 
```
Instrumented module:C:\Software\Tests\UnitTests\CoverletTestingLibraryTests\bin\Debug\net461\NetStandard.Library.31.dll

BadImageFormatException during MetadataReaderProvider.FromPortablePdbStream in InstrumentationHelper.PortablePdbHasLocalSource, unable to check if module has got local source.

Instrumented module:C:\Software\Tests\UnitTests\CoverletTestingLibraryTests\bin\Debug\net461\NetStandard.Library.32.dll

BadImageFormatException during MetadataReaderProvider.FromPortablePdbStream in InstrumentationHelper.PortablePdbHasLocalSource, unable to check if module has got local source.
```

If I run the same command again it will output the same messages but instead of hanging it will continue to completion. The unit tests succeed but it only outputs coverage for 4 of the 5 modules. I assume that is because the previous run attempted to alter the PDBs, somehow failed, and the altered PDBs are no longer valid for one of the projects. I believe that because it claims the PDB is a public PDB when it is not: "Unable to instrument module: C:\Software\Tests\UnitTests\CoverletTestingLibraryTests\bin\Debug\net461\NetFramework.Library.20.dll because : Could not find the '/NAMES' stream: the PDB file may be a public symbol file instead of a private symbol file".

Here is the full log of the second run: 
```
PS C:\Users\cjewett> coverlet C:\Software\Tests\UnitTests\CoverletTestingLibraryTests\bin\Debug\net461\CoverletTestingLibraryTests.dll --target "dotnet" --targetargs "vstest C:\Software\Tests\UnitTests\CoverletTestingLibraryTests\bin\Debug\net461\CoverletTestingLibraryTests.dll" --format cobertura --output "C:/Software" --verbosity detailed

Instrumented module: 'C:\Software\Tests\UnitTests\CoverletTestingLibraryTests\bin\Debug\net461\NetStandard.Library.31.dll'

BadImageFormatException during MetadataReaderProvider.FromPortablePdbStream in InstrumentationHelper.PortablePdbHasLocalSource, unable to check if module has got local source.

Instrumented module: 'C:\Software\Tests\UnitTests\CoverletTestingLibraryTests\bin\Debug\net461\NetStandard.Library.32.dll'

BadImageFormatException during MetadataReaderProvider.FromPortablePdbStream in InstrumentationHelper.PortablePdbHasLocalSource, unable to check if module has got local source.

Unable to instrument module: C:\Software\Tests\UnitTests\CoverletTestingLibraryTests\bin\Debug\net461\NetFramework.Library.20.dll because : Could not find the '/NAMES' stream: the PDB file may be a public symbol file instead of a private symbol file

BadImageFormatException during MetadataReaderProvider.FromPortablePdbStream in InstrumentationHelper.PortablePdbHasLocalSource, unable to check if module has got local source.

Instrumented module: 'C:\Software\Tests\UnitTests\CoverletTestingLibraryTests\bin\Debug\net461\NetStandard.Library.33.dll'

BadImageFormatException during MetadataReaderProvider.FromPortablePdbStream in InstrumentationHelper.PortablePdbHasLocalSource, unable to check if module has got local source.

Instrumented module: 'C:\Software\Tests\UnitTests\CoverletTestingLibraryTests\bin\Debug\net461\NetFramework.Library.21.dll'

Unable to instrument module: C:\Software\Tests\UnitTests\CoverletTestingLibraryTests\bin\Debug\net461\Microsoft.SqlServer.Types.dll because : No symbol found for file: C:\Software\Tests\UnitTests\CoverletTestingLibraryTests\bin\Debug\net461\Microsoft.SqlServer.Types.dll

Unable to instrument module: C:\Software\Tests\UnitTests\CoverletTestingLibraryTests\bin\Debug\net461\protobuf-net.dll, embedded pdb without local source files, [C:\code\protobuf-net\src\protobuf-net\obj\Release\net40\protobuf-net.Version.cs]

Microsoft (R) Test Execution Command Line Tool Version 16.6.0
Copyright (c) Microsoft Corporation.  All rights reserved.
Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
Test Run Successful.
Total tests: 2
     Passed: 2
 Total time: 1.6138 Seconds

Calculating coverage result...
```
Calculating the results worked. I just omitted here. 

The assembly missing code coverage is .NET Framework 20 which is also the one it complains about having a public PDB instead of a private PDB. The reason that happens though is because the previous run froze. Any ideas what might be going on here? Also, what steps can I take to debug this locally and confirm why this happens so I can see if it is possible to replicate easily? 

Thanks for taking a look!

**Comment by Cjewett** [comment]
Further debugging suggests this is an issue with Mono.Cecil. It hangs when ModuleDefinition.Write(Stream stream, WriterParameters parameters) is called for one of the assemblies. Going through the motions to debug that now.

**Comment by Cjewett** [comment]
At some point Cecil.Mono attempts to call ModuleDefinition.Resolve(SqlDbType) and it enters an endless loop that leads to a StackOverflow exception. 

https://github.com/jbevain/cecil/blob/c1a3c199c31b76e3064316a5f24e176ca780b0af/Mono.Cecil/MetadataResolver.cs#L131

The code eventually enters that function.
module = System.Data.dll
reference = System.Data.SqlDbType

It enters the exported_types loop and the reference passed into the function matches one of the exported_types. It then calls exported_types.Resolve() which creates a new TypeReference based on the exported type, and calls module.Resolve(TypeReference) with the newly created TypeReference as the parameter. 

It then repeats the same sequence over and over because one of the exported_types in System.Data.dll always matches the reference (System.Data.SqlDbType) passed into the module.Resolve(TypeReference) call. 

It's possible I should be creating a new issue on Cecil.Mono's Github page but wanted to get thoughts here first.

**Comment by MarcoRossignoli** [comment]
> These PDBs also aren't portable ones

Coverlet works only with portable pdb's

> At some point Cecil.Mono attempts to call ModuleDefinition.Resolve(SqlDbType) and it enters an endless loop that leads to a StackOverflow exception.

There were some similar issue with Cecil in past but we fixed that https://github.com/jbevain/cecil/issues/575
Were related to `netstandard.dll` resolution on core/.NET full.
Which version are you using?

**Comment by Cjewett** [comment]
Updated PDBs to portable and the warnings are gone. Thanks! 

.NET Standard 2.0.3.

The reference tree is .NET Framework 4.6.1 unit test project references .NET Framework 4.6.1 library. The .NET Framework 4.6.1 library references .NET Standard 2.0.3 library. 

I was able to create a reproducible example here: https://github.com/Cjewett/CoverletIssue

**Comment by MarcoRossignoli** [comment]
I think you should move your project to `net472` as reported in netstandard guide there are some issue with `net461` and `netstandard2.0`

https://docs.microsoft.com/en-us/dotnet/standard/net-standard#net-implementation-support 
```
2 The versions listed here represent the rules that NuGet uses to determine whether a given .NET Standard library is applicable. While NuGet considers .NET Framework 4.6.1 as supporting .NET Standard 1.5 through 2.0, there are several issues with consuming .NET Standard libraries that were built for those versions from .NET Framework 4.6.1 projects. For .NET Framework projects that need to use such libraries, we recommend that you upgrade the project to target .NET Framework 4.7.2 or higher.
```
I did a test with your repo project moving to  `net472` and it works
![image](https://user-images.githubusercontent.com/7894084/87228014-ba88d480-c39e-11ea-93fd-f44e2169cab0.png)

**Comment by Cjewett** [comment]
Thanks! I'll test some of upgrade scenarios and report back.

**Comment by Cjewett** [comment]
@MarcoRossignoli So upgrading the test repo did work and I noticed a lot of the assemblies in the output folder for 461 were not in 4.7.2 or 4.8. When upgrading my companies source that was not the case though and it still looped into a stackoverflow exception. I'm trying to figure out reproduce steps but it's difficult because all of our unit test projects are 80+ project solutions. 

This all seems to happen because we have a .NET Core tool running against a .NET Framework unit test project according to the previous issues you linked. Would this be solved if there was a coverlet.console.framework tool, so it was a .NET Framework tool running against a .NET Framework test project? I'm assuming that Mono.Cecil will determine its NET_CORE directive based on the applications build target but maybe I'm wrong there. The reason I ask is I swapped some logic in the Mono.Cecil libraries to use the .NET Framework functionality and it was able to resolve everything correctly.

**Comment by MarcoRossignoli** [comment]
> Would this be solved if there was a coverlet.console.framework tool, so it was a .NET Framework tool running against a .NET Framework test project? 

I don't know if it's possible, I think that coverlet was born primary for .NET core app, we never had issue like your in past so I wonder if there are other solution.

Btw cc: @tonerdo to know what he thinks about

**Comment by Cjewett** [comment]
@MarcoRossignoli  Thanks for continuing this conversation. I've updated https://github.com/Cjewett/CoverletIssue with net48 reproduce steps. SqlDbType seems to resolve correctly now but SqlBulkCopyOptions does not and results in an endless loop. I'm going to mess around with Mono.Cecil and try to figure out how the NET_CORE symbol is set.

**Comment by Cjewett** [comment]
@MarcoRossignoli  I was able to duplicate Coverlet.Console as a .NET Framework tool with some minor adjustments and it ran without hitting an endless loop. The resulting coverage looked okay but I'll need to run it against larger projects.

Cecil.Mono seems to define NET_CORE based on this property: https://github.com/jbevain/cecil/blob/5e37f44535e0753ff3e4dbdef4ef148ca8bb09e0/Directory.Build.props#L10

My expectation is that if you have a .NET Framework project reference Coverlet.Core which references Cecil.Mono the NET_CORE symbol would not be defined. That was not the case though and it was defined. It looks like it will use the underlying libraries target framework rather than the applications if the application does not have a direct reference to the project.

.NET Framework Tool (not referencing Cecil.Mono) that references Coverlet.Core results in NET_CORE defined. NET_CORE is defined based on Coverlet.Core's target framework, not the applications target framework. 

.NET Framework Tool (referencing Cecil.Mono) that references Coverlet.Core results in NET_CORE not defined. NET_CORE is defined based on .NET Framework Tool's target framework, not an underlying libraries target framework.

I imagine the same could be said about the current Coverlet.Console tool but it just so happens the underlying libraries match what you want to use. 

TLDR: It seems like creating a .NET Framework tool that has direct references to Cecil libraries to make sure it defines the NET_CORE symbol based on the application's target framework rather than any underlying libraries will work.

**Comment by Cjewett** [comment]
@MarcoRossignoli  Thoughts on the previous reply? I'm likely to fork and package a .NET Framework version of the tool for my company to use but it would be nice if it was just baked in to the project directly.

**Comment by MarcoRossignoli** [comment]
I don't think we'll ship a version specific for .NET Framework, we're using .NET standard refs to avoid this.
In past we had a similar issue, so I would debug your new sample with `SqlBulkCopyOptions` loops to understand where is the problem(seem something like past issue where a wrong netstandard lib is loaded and type is not found so retry forever).
Can you confirm that https://github.com/Cjewett/CoverletIssue repro new `SqlBulkCopyOptions` loop?So I can give it a try

**Comment by Cjewett** [comment]
@MarcoRossignoli  Yes, that project has the repro.

**Comment by MarcoRossignoli** [comment]
close for stale conversation, feel free to reopen if needed.
